lib/chef/resource/sysctl.rb

```diff
@@ -0,0 +1,123 @@
+#
+# Copyright:: 2018, Webb Agile Solutions Ltd.
+# Copyright:: 2018, Chef Software Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "chef/resource"
+
+class Chef
+  class Resource
+    class Sysctl < Chef::Resource
+      resource_name :sysctl
+      provides :sysctl
+      provides :sysctl_param
+
+      description "Use the sysctl resource to set kernel parameters using the sysctl"\
+                  " command line tool and configuration files in the system's sysctl.d directory."\
+                  "Configuration files managed by this resource are named 99-chef-KEYNAME.conf. If"\
+                  " an existing value was already set for the value it will be backed up to the node"\
+                  " and restored if the :remove action is used later."
+
+      introduced "14.0"
+
+      property :key, String,
+               description: "",
+               name_property: true
+
+      property :ignore_error, [TrueClass, FalseClass],
+               description: "",
+               default: false
+
+      property :value, [Array, String, Integer, Float],
+               description: "",
+               coerce: proc { |v| coerce_value(v) },
+               required: true
+
+      property :conf_dir, String,
+               description: "",
+               default: "/etc/sysctl.d"
+
+      def after_created
+        raise "The systctl resource requires Linux as it needs sysctl and the systctl.d directory functionality." unless node["os"] == "linux"
+        raise "The systctl resource does not support SLES releases less than 12 as it requires a systctl.d directory" if platform_family?("suse") && node["platform_version"].to_i < 12
+      end
+
+      def coerce_value(v)
+        case v
+        when Array
+          v.join(" ")
+        else
+          v.to_s
+        end
+      end
+
+      def get_sysctl_value(key)
+        o = shell_out("sysctl -n -e #{key}")
+        raise "Unknown sysctl key #{key}!" if o.error!
```

Comment:
`error!` raises. so this is ... a bit meaningless :) did you mean `error?`

Reply:
Oh nice that never worked in the cookbook either. I'll get that fixed